kernel/src/engine/ensure_data_types.rs

```diff
@@ -0,0 +1,486 @@
+/// asking for exist, and that for those fields the types match. Un-selected fields are ignored.
+pub(crate) fn ensure_data_types(
+    kernel_type: &DataType,
+    arrow_type: &ArrowDataType,
+    check_nullability_and_metadata: bool,
+) -> DeltaResult<DataTypeCompat> {
+    let check = EnsureDataTypes { check_nullability_and_metadata };
+    check.ensure_data_types(kernel_type, arrow_type)
+}
+
+struct EnsureDataTypes {
+    check_nullability_and_metadata: bool,
+}
+
+/// Capture the compatibility between two data-types, as passed to [`ensure_data_types`]
+pub(crate) enum DataTypeCompat {
+    /// The two types are the same
+    Identical,
+    /// What is read from parquet needs to be cast to the associated type
+    NeedsCast(ArrowDataType),
+    /// Types are compatible, but are nested types. This is used when comparing types where casting
+    /// is not desired (i.e. in the expression evaluator)
+    Nested,
+}
+
+impl EnsureDataTypes {
+    // Perform the check. See documentation for `ensure_data_types` entry point method above
+    fn ensure_data_types(
+        &self,
+        kernel_type: &DataType,
+        arrow_type: &ArrowDataType,
+    ) -> DeltaResult<DataTypeCompat> {
+        match (kernel_type, arrow_type) {
+            (DataType::Primitive(_), _) if arrow_type.is_primitive() => {
+                check_cast_compat(kernel_type.try_into()?, arrow_type)
+            }
+            (&DataType::BOOLEAN, ArrowDataType::Boolean)
+                | (&DataType::STRING, ArrowDataType::Utf8)
+                | (&DataType::BINARY, ArrowDataType::Binary) => {
+                    // strings, bools, and binary  aren't primitive in arrow
+                    Ok(DataTypeCompat::Identical)
+                }
+            (DataType::Array(inner_type), ArrowDataType::List(arrow_list_field)) => {
+                self.ensure_nullability(
+                    "List",
+                    inner_type.contains_null,
+                    arrow_list_field.is_nullable(),
+                )?;
+                let kernel_array_type = &inner_type.element_type;
+                let arrow_list_type = arrow_list_field.data_type();
+                self.ensure_data_types(
+                    kernel_array_type,
+                    arrow_list_type,
+                )
+            }
+            (DataType::Map(kernel_map_type), ArrowDataType::Map(arrow_map_type, _)) => {
+                if let ArrowDataType::Struct(fields) = arrow_map_type.data_type() {
+                    let mut fields = fields.iter();
+                    if let Some(key_type) = fields.next() {
+                        self.ensure_data_types(
+                            &kernel_map_type.key_type,
+                            key_type.data_type(),
+                        )?;
+                    } else {
+                        return Err(make_arrow_error(
+                            "Arrow map struct didn't have a key type".to_string(),
```

Comment:
We shouldn't need these `.to_string()` calls any more?
```suggestion
                            "Arrow map struct didn't have a key type",
```
(more below)